ui-v2/tests/unit/utils/get-component-factory-test.js

```diff
@@ -0,0 +1,48 @@
+import getComponentFactory from 'consul-ui/utils/get-component-factory';
+import { module, test } from 'qunit';
+
+module('Unit | Utility | get component factory');
+
+test("it uses lookup to locate the instance of the component based on the DOM element's id", function(assert) {
+  const expected = 'name';
+  let getComponent = getComponentFactory({
+    lookup: function() {
+      return { id: expected };
+    },
+  });
+  assert.equal(typeof getComponent, 'function', 'returns a function');
+  const actual = getComponent({
+    getAttribute: function(name) {
+      return 'id';
+    },
+  });
+  assert.equal(actual, expected, 'performs a lookup based on the id');
+});
+test("it returns null if it can't find it", function(assert) {
+  const expected = null;
+  let getComponent = getComponentFactory({
+    lookup: function() {
+      return { id: '' };
+    },
+  });
+  const actual = getComponent({
+    getAttribute: function(name) {
+      return 'non-existent';
+    },
+  });
+  assert.equal(actual, expected);
+});
+test('it returns null if there is no id', function(assert) {
+  const expected = null;
+  let getComponent = getComponentFactory({
+    lookup: function() {
+      return { id: '' };
+    },
+  });
+  const actual = getComponent({
+    getAttribute: function(name) {
+      return;
+    },
+  });
+  assert.equal(actual, expected);
+});
```

Comment:
These tests look fine, but the concept of a `getComponentFactory` is an anti-pattern. You shouldn't be looking up views and invoking things on them from controllers. That's the antithesis of data-down actions-up.

Reply:
So this is a bit of a bigger one. I wouldn't necessarily say this is an 'anti-pattern'. I had a good think about this before doing it this way, and I read plenty of opinions online by ember-y people who also thought that DDAU wasn't the ideal way to deal with this either. I'd possibly equate it to saying that you should use DDAU for window resizing instead of listening to the resize event, which probably isn't a good plan.

Other options I read about we concerned with making a global message bus, or registering the components with the parent controller via attributes, both of which weren't decoupled enough for me. I didn't want to add anything special to the component in order to achieve what I needed.

What I've done is pretty much how standard web-components work, well.. as close to it as I could get. Happy to discuss this more if you feel it is absolutely the wrong thing to do.